docs/design_proposals/sync-auto-support.md

```diff
@@ -0,0 +1,150 @@
+  - This is similar to `_smart_` described in [sync-improvements](sync-improvements.md)
+
+3. Is there any another workaround, and if so, what are its drawbacks?
+  - Currently one can create a docker file that only copies specific build
+    results into the container and relies on a local build to generate those
+    intermediate artifacts. This requires a user to trigger a first build
+    manually before docker kicks in to containerize the application. While
+    it may be possible to automate this (for example: gradle --continuous), it
+    is not an acceptable solution to require manual external processes for
+    a build to succeed.
+
+4. Mention related issues, if there are any.
+  - This is not trying to solve the problem of dealing with a multistage
+    dockerbuild. Intermediate build artifacts might still be possible to
+    determine, however that would require an extra mechanism to do a partial
+    docker build and sync files from a built container to a running container --
+    something we do not intend to cover here.
+
+#### Problems with current API/config
+The current `sync` system has the following problems:
+1. No way to trigger local external processes - for example, in jib, to build a
+   container, one would run `./gradlew jib`, but to update class files so the
+   system may sync them, one would only be required to run `./gradlew classes`.
+2. No way to tell the system to sync non build inputs. Skaffold is
+   normally only watching `.java` files for a build, in the sync case, we want
+   it to watch `.java` files, trigger a partial build, and sync `.class` files
+   so a remote server can pick up and reload the changes.
+
+## Design
+
+#### Hack it
+To get close to the functionality we want, without modifying skaffold at all, a
+Dockerfile which depends on java build outputs could be used, like:
+```
+FROM openjdk:8
+COPY build/dependencies/ /app/dependencies
+COPY build/classes/java/main/ /app/classes
+
+CMD ["java", "-cp", "/app/classes:/app/dependencies/*", "hello.Application"]
+```
+
+with a skaffold sync block that looks like:
+```
+sync:
+  manual:
+    - src: "build/classes/java/main/**/*.class"
+      dest: "/app/classes"
+      strip: "build/classes/java/main/"
+```
+
+A user's devloop then looks like this:
+
+1. run `./gradlew classes copyDependencies`
+2. run `skaffold dev`
+3. *make changes to some java file*
+4. run `./gradlew classes`
+5. *skaffold syncs files*
+
+which is far from ideal.
+
+#### Fix it?
+
+What we might want, is an API that can accept configuration that looks like:
+
+```
+sync: auto {
+  indirect:
+  - command: "./gradlew jib classes"
+    inputs:
+      - src/main/java
+      - src/main/resources
+    files:
+      - src: "target/classes"
+        dest: "/app/classes",
+      - src: "target/resources.
+        dest: "/app/resources",
+  - command: "compile_c_stuff.sh"
+    inputs:
+    - asdf.c
+    files:
+     - src: "asdf.o"
+       dest: "/libs/asdf.o"
+  direct:
```

Comment:
Maybe I haven't thought enough about the edge cases. But is it really necessary to separate out the `direct` sync cases? What would be wrong about a sync rule that applies to both generated artifacts and direct source files?

Reply:
I think the process of doing a build when it shouldn't could be costly?
And if the build system isn't smart, it could trigger sync'ing of files that didn't need to be (and this could cause a server reload when there didn't need to be?)

Reply:
My understanding was that the build hook is only triggered, if one of its input files changed (and Skaffold takes care to detect this). The syncing could be considered as a different step, where everything that changed is sync'ed. And if the build hook didn't run, none of its outputs would have changed. Or is that not how you planned it?

Reply:
Oh I see. I planned for it all to exist in the `auto` block in this example as a single sync rule. Naming is a little weird at the moment.